tools/SwiftSyntax/SyntaxData.swift

```diff
@@ -38,6 +38,79 @@ final class SyntaxData: Equatable {
 
   let childCaches: [AtomicCache<SyntaxData>]
 
+  let positionCache: AtomicCache<AbsolutePosition>
+
+  let positionAfterSkippingLeadingTriviaCache: AtomicCache<AbsolutePosition>
+
+  let byteSizeCache: AtomicCache<ByteSize>
+
+  fileprivate func calculatePosition() -> AbsolutePosition {
+    guard let parent = parent else {
+      let message = "cannot find SourceFileSyntax as root"
+      switch raw {
+      case .node(let kind, _, _):
+        assert(kind == SyntaxKind.sourceFile, message)
+      default:
+        assert(false, message)
+      }
+      // If this node is SourceFileSyntax, its location is the start of the file.
+      return AbsolutePosition()
+    }
+
+    // If the node is the first child of its parent, the location is same with
+    // the parent's location.
+    guard indexInParent != 0 else { return parent.position }
+
+    // Otherwise, the location is same with the previous sibling's location
+    // adding the stride of the sibling.
+    for idx in (0..<indexInParent).reversed() {
+      if let sibling = parent.cachedChild(at: idx) {
+        let pos = sibling.position
```

Comment:
My biggest concern here is this recursion. If a function has many statements, this recursion might cause stack-overflow.

Reply:
I've added a test in the following commit that addresses review comments. The test created a statement list of two thousand statements and tried to get the position of the last element. It seems to work fine. I think in practice the stack overflow may be very unlikely to happen.